src/components/Button/v1/Button.js

```diff
@@ -1,50 +1,83 @@
 import React, { Component } from "react";
 import PropTypes from "prop-types";
 import styled from "styled-components";
-import { applyTheme, preventAccidentalDoubleClick } from "helpers";
+import { applyTheme, preventAccidentalDoubleClick, spinner } from "helpers";
 import { applyThemeWithActionType } from "./helpers";
 
 const paddingFunc = (props) => {
-  const { styleVariant } = props;
-  if (styleVariant === "small") return applyTheme("buttonVerticalPaddingSmall");
+  const { isShortHeight } = props;
+  if (isShortHeight) return applyTheme("buttonVerticalPaddingShort");
   return applyTheme("buttonVerticalPadding");
 };
 
 const ButtonDiv = styled.div`
-  background-color: ${applyThemeWithActionType("buttonBackgroundColor", true)};
-  border-color: ${applyThemeWithActionType("buttonBorderColor", true)};
+  align-items: center;
+  background-color: ${applyThemeWithActionType("buttonBackgroundColor")};
+  border-color: ${applyThemeWithActionType("buttonBorderColor")};
   border-style: solid;
   border-width: 1px;
   border-radius: ${applyTheme("buttonBorderRadius")};
   box-sizing: border-box;
-  color: ${applyThemeWithActionType("buttonForegroundColor", true)};
+  color: ${applyThemeWithActionType("buttonForegroundColor")};
   cursor: pointer;
   display: ${(props) => {
-    const { children, fullWidth } = props;
+    const { fullWidth } = props;
     if (fullWidth) {
-      if (typeof children === "string") return "block";
       return "flex";
     }
-    if (typeof children === "string") return "inline-block";
     return "inline-flex";
   }};
+  justify-content: center;
   margin: 0;
   min-width: ${applyTheme("buttonMinimumWidth")};
   outline: none;
   padding-left: ${applyTheme("buttonHorizontalPadding")};
   padding-right: ${applyTheme("buttonHorizontalPadding")};
   padding-top: ${paddingFunc};
   padding-bottom: ${paddingFunc};
+  position: relative;
   text-align: center;
+
+  &:hover {
+    background-color: ${applyThemeWithActionType("buttonBackgroundColor", "hover")};
+    border-color: ${applyThemeWithActionType("buttonBorderColor", "hover")};
+    color: ${applyThemeWithActionType("buttonForegroundColor", "hover")};
+  }
+
+  &:active {
+    background-color: ${applyThemeWithActionType("buttonBackgroundColor", "active")};
+    border-color: ${applyThemeWithActionType("buttonBorderColor", "active")};
+    color: ${applyThemeWithActionType("buttonForegroundColor", "active")};
+  }
+`;
+
+const SpinnerWrap = styled.div`
+  display: flex;
+  overflow: hidden;
+  padding-left: ${applyTheme("buttonHorizontalPadding")};
+  transition: width 0.2s ease-out 0s, padding-left 0.2s ease-out 0s, opacity 0.2s ease-out 0.2s;
```

Comment:
I normally wouldn't animate on `width` or `padding` properties because of these properties not using the GPU while animating. I ran the animation with the performance profile running and the animation is getting around 25-35fps. I tried to promote it to a composite layer by adding `will-change` and then `translate3d` but the fps was about the same.

That all being said I can't really think of a way to achieve this animation style without using those properties within the transition.

Reply:
What's the problem too much bandwidth? If it's a major issue we can try and think of another way to do it.